contracts/Governance/GovernorAlpha.sol

```diff
@@ -329,3 +329,50 @@ interface TimelockInterface {
 interface CompInterface {
     function getPriorVotes(address account, uint blockNumber) external view returns (uint96);
 }
+
+contract QuadraticGovernorAlpha is GovernorAlpha {
+    /// @notice The name of this contract
+    string public constant name = "Quadratic Governor Alpha";
+
+    function sqrt(uint y) internal pure returns (uint z) {
+        if (y > 3) {
+            z = y;
+            uint x = y / 2 + 1;
+            while (x < z) {
+                z = x;
+                x = (y / x + x) / 2;
+            }
+        } else if (y != 0) {
+            z = 1;
+        }
+    }
+
+    function _castVote(address voter, uint proposalId, bool support) internal {
+        /* insert POH code */
+        IProofOfHumanity POH = IProofOfHumanity(0xC5E9dDebb09Cd64DfaCab4011A0D5cEDaf7c9BDb);
+        require (POH.isRegistered(voter), "Voter must be registered on POH");
+        /* end POH code */
+
+        require(state(proposalId) == ProposalState.Active, "GovernorAlpha::_castVote: voting is closed");
+        Proposal storage proposal = proposals[proposalId];
+        Receipt storage receipt = proposal.receipts[voter];
+        require(receipt.hasVoted == false, "GovernorAlpha::_castVote: voter already voted");
+        uint96 votes = comp.getPriorVotes(voter, proposal.startBlock);
+        /* insert quadratic votes */
+        votes = sqrt(votes);
```

Comment:
to achieve [quadratic liquid democracy](https://twitter.com/owocki/status/1347915959487524865), we would probably want to square root each of the delegated votes individually

unless we want to squash people who have a broad distribution of support (i dont think we do)

Reply:
i think we want to make it so you collect sum(sqrt(x) for x in delegates) instead of sqrt(sum(x for x in delegates))

that way youre collecting humnans who delegate to you instead of collecting whales

Reply:
I am interested in working on this, or at least thinking about it. It took me a while to wrap my head around sqrt so I made a [spreadsheet](https://docs.google.com/spreadsheets/d/1dPZv7Map9R7Z7d5Vi0Efi-z2AMINXj0ianIATd7K45A/edit?usp=sharing) to try to capture the intent of quadratic voting applied to this delegate model. I think it brought me to the same conclusion here if x would be the number of tokens contributed by each member represented by the delegate. If I am understanding correctly, that complicates the code here a bit. If a delegate has been delegated votes by three token holders in the amount of 1, 4, 9, then I think the votes allotted to the delegate would be sqrt(1) + sqrt(4) + sqrt(9). If I am not off here, I think you would need an array of votes/tokens allotted to the delegates, or a new method added to `CompInterface` that handles that math.

Reply:
>  I am understanding correctly, that complicates the code here a bi

yeah it does unfortunately.

@androolloyd and i have been jamming about making, since this is so computationally complex (and there are challenges in that its hard to know the token balances / delegates to/from in an elegant way), so we are considering computing it off-chain or in a L2 and having an oracle go back to L1 .

happy to share design notes if you are serious about colalborating here !

Reply:
I would be interested, at least to try to wrap my head around the problem. 

Looked at the code a bit that @androolyd linked. It seems like in Quad, the use of sqrt96 would do the trick, but I think I see what your getting at with the complexity, the accounting could get unwieldy with token transfers. I think it could work if you consider the tokens before and after the transfer. Something like `voteDelta = sqrt(tokensAfter) - sqrt(tokensBefore)` So if someone had 4 tokens and they were delegated as 2 votes, and then they received 5 tokens then the `voteDelta = sqrt(9) - sqrt(4) =  1` then add the `voteDelta` to the balance of the delegetee? Just throwing things at the wall to see if they stick, so maybe I should review your notes in case I am going over ground already covered.

Reply:
i think that we need to disambiguate vote credits from votes.  the token balance = vote credits, the sqrt(token_balance) is the vote amount.

i'm not exactly sure what youre getting at with the voteDelta thing here.   but yes there is additional complexity introduced by this all being in GTC.sol instead of in the governoralpha, and also the need to use checkpoint'd token balances, and the need to calculate the votes from the vote credits for every single instance of `_castVote`.

happy to hop on. acall the 3 of us if it helps

Reply:
How the GovernorAlpha is designed is that all balances and votes are computed/stored in the token's state, 

in the token design the checkpoints is your vote amount, and the balances are your vote credits. 

Squaring the data before adding to the checkpoints gives you quadratic voting, but it encapsulates all the functionality into the token, so absent migrating the token to a new token, we are always going to be stuck with what our token can do, which is currently vote credits == votes. 

Computing this off chain on an L2, like Truebit or something of that effect and verifying its correctness on chain. 

This would be huge as it lets the current token design remain and for the governance structures to be updated/iterated on.